psutil/__init__.py

```diff
@@ -888,33 +888,33 @@ def children(self, recursive=False):
                     except (NoSuchProcess, ZombieProcess):
                         pass
         else:
-            # construct a dict where 'values' are all the processes
-            # having 'key' as their parent
-            table = collections.defaultdict(list)
+            # Construct a {pid: [child pids]} dict
+            reverse_ppid_map = collections.defaultdict(list)
             for pid, ppid in ppid_map.items():
-                try:
-                    p = Process(pid)
-                    table[ppid].append(p)
-                except (NoSuchProcess, ZombieProcess):
-                    pass
-            # At this point we have a mapping table where table[self.pid]
-            # are the current process' children.
-            # Below, we look for all descendants recursively, similarly
-            # to a recursive function call.
-            checkpids = [self.pid]
-            for pid in checkpids:
-                for child in table[pid]:
+                reverse_ppid_map[ppid].append(pid)
+            # Recursively traverse that dict, starting from self.pid,
+            # such that we only call Process() on actual children
+            seen = set()
+            stack = [self.pid]
+            while stack:
+                pid = stack.pop()
+                if pid in seen:
+                    # Since pids can be reused while the ppid_map is
```

Comment:
Note I'm not sure this check is necessary, as the `create_time` test may be sufficient to avoid such potential cycles.